docs/design.md

```diff
@@ -0,0 +1,431 @@
+* `substitutions` - list, optional - A sequence of substitutions to apply. See
+  the Substitutions section for additional details.
+  * `dest` - dict, required - A description of the inserted content destination.
+    * `path` - string, required - The JSON path where the data will be placed
+      into the `data` section of this document.
+  * `src` - dict, required - A description of the inserted content source.
+    * `schema` - string, required - The `schema` of the source document.
+    * `name` - string, required - The `metadata.name` of the source document.
+    * `path` - string, required - The JSON path from which to extract data in
+      the source document relative to its `data` section.
+
+Here is a fictitious example of a complete document which illustrates all the
+valid fields in the `metadata` section.
+
+```yaml
+---
+schema: some-service/ResourceType/v1
+metadata:
+  schema: metadata/Document/v1
+  name: unique-name-given-schema
+  storagePolicy: cleartext
+  labels:
+    genesis: enabled
+    master: enabled
+  layeringDefinition:
+    abstract: true
+    layer: region
+    parentSelector:
+      labels: parents
+      must: have
+    actions:
+      - method: merge
+        path: .path.to.merge.into.parent
+      - method: delete
+        path: .path.to.delete
+  substitutions:
+    - dest:
+        path: .substitution.target
+      src:
+        schema: another-service/SourceType/v1
+        name: name-of-source-document
+        path: .source.path
+data:
+  path:
+    to:
+      merge:
+        into:
+          parent:
+            foo: bar
+          ignored:  # Will not be part of the resultant document after layering.
+            data: here
+  substitution:
+    target: null  # Paths do not need to exist to be specified as substitutiond destinations.
+```
+
+
+##### schema: metadata/Control/v1
+
+This schema is the same as the `Document` schema, except it omits the
+`storagePolicy`, `layeringDefinition`, and `substitutions` keys, as these
+actions are not supported on `Control` documents.
+
+The complete list of valid `Control` document kinds is specified below along
+with descriptions of each document kind.
+
+##### schema: metadata/Tombstone/v1
+
+The only valid key in a `Tombstone` metadata section is `name`.  Additionally,
+the top-level `data` section should be omitted.
+
+### Layering
+
+When queried, Deckhand assembles the chain of partial documents in the order
+specified in the `deckhand/LayeringPolicy/v1` document. Each chain of partial
+documents is made up only of documents using a single `schema`.
+
+<!-- TODO: describe document chain construction w/ diagrams -->
+
+During each layering step, the list of `actions` will be applied in order.
+Supported actions are:
+
+* `merge` - a "deep" merge that layers new and modified data onto existing data
+* `replace` - overwrite data at the specified path and replace it with the data
+  given in this document
+* `delete` - remove the data at the specified path
+
```

Comment:
After some more conversation, the direction we are looking at is to only allow a single parent at each layer, which avoids some of the concerns here.